docs/architecture.md

```diff
@@ -0,0 +1,221 @@
+   - relay the response back to the client
+ 7. release processing pipeline serialization lock
+
+The high-level control flow of the event processing pipeline is one of the
+following, based on the event type:
+
+ - for policy-specific events:
+   1. take processing pipeline lock
+   2. call policy event handler
+   3. invoke the controller layer for post-policy processing (same as step 5 for requests)
+   4. release pipeline lock
+ - for metrics events:
+   1. perform collection/processing/correlation
+   2. take processing pipeline lock
+   3. update cache objects as/if necessary
+   4. request rebalancing as/if necessary
+   5. release pipeline lock
+ - for rebalance events:
+   1. take processing pipeline lock
+   2. invoke policy layer for rebalancing
+   3. invoke the controller layer for post-policy processing (same as step 5 for requests)
+   4. release pipeline lock
+
+
+### Cache
+
+The cache is a shared internal storage within CRI-RM. It tracks the runtime
+state of pods and containers known to CRI-RM, as well as the state of CRI-RM
+itself, including the active configuration and the state of the active policy.
+The cache is saved to permanent storage in the filesystem and used to restore
+the runtime state of CRI-RM across restarts.
+
+The cache provides functions for querying and updating the state of pods and
+containers. This is the mechanism used by the active policy to make resource
+assignment decisions. The policy simply updates the state of the affected
+containers in the cache according to the decisions.
+
+The cache's ability to associate and track changes to containers with resource
+domains is used to enforce policy decisions. The generic controller layer first
+queries which containers have pending changes, then invokes each controller for
+each containers. The controllers use the querying functions provided by the
+cache to decide if anything in their resource/control domain needs to be changed
+and then act accordingly.
+
+Access to the cache needs to be serialized. However, this serialization is not
+provided by the cache itself. Instead, it assumes callers to make sure proper
+protection is in place against concurrent read-write access. The request and
+event processing pipelines in the resource manager use a lock to serialize
+request and event processing and consequently access to the cache.
+
+
+### Generic Policy Layer
+
+The generic policy layer defines the abstract interface the rest of CRI-RM
+uses to interact with policy implementations and takes care of the details
+of activating and dispatching calls through to the configured active policy.
+
+
+### Generic Resource Controller Layer
+
+The generic resource controller layer defines the abstract interface the rest
+of CRI-RM uses to interact with resource controller implementations and takes
+care of the details of dispatching calls to the controller implementations for
+post-policy enforcment of decisions.
+
+
+### Metrics Collector
+
```

Comment:
Some text here 😉  Should we mention tracing, too?

Reply:
Sure. It's just that metrics collection is still a bit in flux/WiP. We've been a little too busy with other stuff to really push it to a level where it'd definitely start taking a final shape.

Reply:
Yes, tracing is missing, as well as the introspection+visualisation.

Reply:
Wrote a paragraph about metrics collection.